pkg/webhook/server/certificate/secret.go

```diff
@@ -12,33 +14,53 @@ import (
 	"github.com/qinqon/kube-admission-webhook/pkg/webhook/server/certificate/triple"
 )
 
-func tlsSecret(service types.NamespacedName, keyPair *triple.KeyPair) corev1.Secret {
+func updateTLSSecret(secret corev1.Secret, keyPair *triple.KeyPair) *corev1.Secret {
+	secret.Data = map[string][]byte{
+		corev1.TLSCertKey:       triple.EncodeCertPEM(keyPair.Cert),
+		corev1.TLSPrivateKeyKey: triple.EncodePrivateKeyPEM(keyPair.Key),
+	}
+	secret.Type = corev1.SecretTypeTLS
+	return &secret
+}
+
+func (m *Manager) newTLSSecret(serviceKey types.NamespacedName, keyPair *triple.KeyPair) (*corev1.Secret, error) {
+	service := corev1.Service{}
+	err := m.get(serviceKey, &service)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed getting service %s to set secret owner", serviceKey)
+	}
 	secret := corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      service.Name,
 			Namespace: service.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					Name:       service.Name,
+					Kind:       service.TypeMeta.Kind,
+					APIVersion: service.TypeMeta.APIVersion,
+					UID:        service.UID},
+			},
 		},
-		Data: map[string][]byte{
-			corev1.TLSCertKey:       triple.EncodeCertPEM(keyPair.Cert),
-			corev1.TLSPrivateKeyKey: triple.EncodePrivateKeyPEM(keyPair.Key),
-		},
-		Type: corev1.SecretTypeTLS,
 	}
-	return secret
+	return updateTLSSecret(secret, keyPair), nil
 }
 
 func (m *Manager) applyTLSSecret(service types.NamespacedName, keyPair *triple.KeyPair) error {
-	tlsSecret := tlsSecret(service, keyPair)
+	secret := corev1.Secret{}
 
-	err := m.get(service, &corev1.Secret{})
+	err := m.get(service, &secret)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			return m.client.Create(context.TODO(), &tlsSecret)
+			tlsSecret, err := m.newTLSSecret(service, keyPair)
+			if err != nil {
+				return errors.Wrapf(err, "failed initailizing secret %s", service)
+			}
+			return m.client.Create(context.TODO(), tlsSecret)
 		} else {
 			return err
 		}
 	}
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
```

Comment:
[just asking to understand) if there's a conflict, don't we need to do an additional get in order to refresh the conflicted object (in our case, secret/keypair)?

Reply:
Ohhh man you are soo right, good catch !